File: python/engine/streamer.py
"""Streaming transcription — sentence-based sliding window + LocalAgreement."""
import re
import time

# Common Whisper hallucinations on silence/noise
HALLUCINATIONS = {
    "", "you", "thank you.", "thanks for watching!", "thanks for watching.",
    "subscribe", "bye.", "bye", "thank you", "you.", "the end.",
    "thanks for listening.", "see you next time.", "thank you for watching.",
    "...", "MBC 뉴스 , 이덕영입니다.", "字幕by索兰娅梦", "请不吝点赞 订阅 转发 打赏支持明镜与点点栏目",
}

# Punctuation that triggers LOCK (sentence-ending + clause-ending)
SENTENCE_END = re.compile(r'[.!?。！？,，;；:：、]\s*$')
SENTENCE_SPLIT = re.compile(r'(?<=[.!?。！？,，;；:：、])\s*')

# Sliding window config
MAX_WINDOW_SECONDS = 12.0  # max audio to re-transcribe (~1.5 sentences)


def _is_cjk(char):
    """Check if a character is CJK (Chinese/Japanese/Korean)."""
    cp = ord(char)
    return (0x4E00 <= cp <= 0x9FFF or 0x3400 <= cp <= 0x4DBF or
            0x3040 <= cp <= 0x309F or 0x30A0 <= cp <= 0x30FF or
            0xAC00 <= cp <= 0xD7AF or 0xFF00 <= cp <= 0xFFEF)


def _tokenize(text):
    """Split text into tokens. CJK: per-character. Others: by whitespace."""
    tokens = []
    buf = []
    for ch in text:
        if _is_cjk(ch):
            if buf:
                tokens.append("".join(buf))
                buf = []
            tokens.append(ch)
        elif ch.isspace():
            if buf:
                tokens.append("".join(buf))
                buf = []
        else:
            buf.append(ch)
    if buf:
        tokens.append("".join(buf))
    return tokens


def _join_tokens(tokens):
    """Join tokens back into text. No spaces between adjacent CJK characters."""
    if not tokens:
        return ""
    parts = [tokens[0]]
    for i in range(1, len(tokens)):
        prev, cur = tokens[i - 1], tokens[i]
        # Add space unless both sides are CJK
        if prev and cur and _is_cjk(prev[-1]) and _is_cjk(cur[0]):
            parts.append(cur)
        else:
            parts.append(" " + cur)
    return "".join(parts)


def _common_prefix_words(words_a, words_b):
    """Longest common prefix between two word lists."""
    length = min(len(words_a), len(words_b))
    for i in range(length):
        if words_a[i] != words_b[i]:
            return words_a[:i]
    return words_a[:length]


def _fuzzy_agree(words_a, words_b, threshold=0.8):
    """Fuzzy agreement: if 80%+ tokens match positionally, return agreed tokens.

    Walks both lists simultaneously. Matching tokens are kept.
    Mismatches: take the token from the LATEST run (words_b).
    Returns (agreed_tokens, match_ratio) or ([], 0.0) if below threshold.
    """
    length = min(len(words_a), len(words_b))
    if length == 0:
        return [], 0.0

    matches = 0
    agreed = []
    for i in range(length):
        if words_a[i] == words_b[i]:
            matches += 1
            agreed.append(words_b[i])
        else:
            agreed.append(words_b[i])  # prefer latest

    ratio = matches / length
    if ratio >= threshold:
        return agreed, ratio
    return [], ratio


def _extract_complete_sentences(text):
    """Split text into (complete_sentences, remainder).

    complete_sentences: text up to and including the last sentence-ending punctuation
    remainder: text after that (the in-progress sentence)
    """
    # Find last sentence-ending punctuation
    parts = SENTENCE_SPLIT.split(text)
    if len(parts) <= 1:
        # Check if the single part ends with punctuation
        if SENTENCE_END.search(text):
            return text, ""
        return "", text

    # Everything except the last part is complete
    # But check if the last part also ends with punctuation
    if SENTENCE_END.search(parts[-1]):
        return text, ""

    complete = " ".join(parts[:-1])
    remainder = parts[-1]
    return complete, remainder


class StreamingTranscriber:
    """Sentence-based sliding window transcription.

    Strategy:
        1. Record continuously, transcribe rolling buffer every N seconds
        2. Word-level LocalAgreement: only trust words that match across 2 runs
        3. When a complete sentence is confirmed → LOCK it, paste it, slide window
        4. Re-transcribe only the audio after the locked sentence (~5-10s)
        5. Final pass: re-transcribe last incomplete sentence for accuracy

    This keeps the buffer short (~10s max) so transcription stays fast (~0.5s),
    and pasting is sentence-by-sentence (clean, no fragile char-count replacement).
    """

    def __init__(self, backend, interval=3.0, sr=16000):
        self.backend = backend
        self.interval = interval
        self.sr = sr

        self._prev_words = []           # words from previous transcription
        self._confirmed_words = []      # words confirmed by agreement

        self._locked_sentences = []     # complete sentences, pasted and done
        self._locked_text = ""          # joined locked sentences
        self._pending_text = ""         # current in-progress sentence (not yet pasted)

        self._window_start_sample = 0   # audio sample index where current window starts

        # Fail safe: force push after N transcriptions of same window
        self._window_pass_count = 0     # how many times current window has been transcribed
        self._max_passes = 3            # force push after this many passes

    def rolling_transcribe(self, full_audio, language=None):
        """Run one rolling transcription pass with fuzzy agreement + fail safe.

        Returns:
            (new_sentence: str or None, pending_text: str)
        """
        window_audio = full_audio[self._window_start_sample:]

        if len(window_audio) < int(0.5 * self.sr):
            return None, self._pending_text

        t0 = time.time()
        text = self.backend.transcribe(window_audio, sr=self.sr, language=language)
        elapsed = time.time() - t0

        if text.lower().strip() in HALLUCINATIONS:
            return None, self._pending_text

        current_words = _tokenize(text)
        self._window_pass_count += 1

        # --- Agreement logic ---
        agreed_words = []

        if self._prev_words:
            # Try strict prefix first
            strict = _common_prefix_words(self._prev_words, current_words)
            if len(strict) >= 3:
                agreed_words = list(strict)
            else:
                # Try fuzzy agreement (80% threshold)
                fuzzy, ratio = _fuzzy_agree(self._prev_words, current_words, threshold=0.8)
                if fuzzy:
                    agreed_words = fuzzy

        self._prev_words = current_words

        # Update confirmed words
        if agreed_words and len(agreed_words) > len(self._confirmed_words):
            self._confirmed_words = agreed_words

        confirmed_text = _join_tokens(self._confirmed_words)

        # --- Fail safe: force push after N passes on same window ---
        if self._window_pass_count >= self._max_passes:
            full_text = _join_tokens(current_words)
            if full_text.strip():
                # Push up to last punctuation, keep the rest for next window
                complete, remainder = _extract_complete_sentences(full_text)
                if complete:
                    self._do_lock(complete, current_words, window_audio, remainder)
                    return complete, self._pending_text
                else:
                    # No punctuation at all — push everything as last resort
                    self._do_lock(full_text, current_words, window_audio)
                    return full_text, self._pending_text

        # --- Normal lock: agreement + punctuation or length ---
        new_sentence = None

        if confirmed_text:
            complete, remainder = _extract_complete_sentences(confirmed_text)

            if complete:
                new_sentence = complete
                self._do_lock(complete, current_words, window_audio, remainder)
            elif len(confirmed_text) > 40:
                new_sentence = confirmed_text
                self._do_lock(confirmed_text, current_words, window_audio)
            else:
                self._pending_text = confirmed_text
        else:
            self._pending_text = _join_tokens(current_words)

        return new_sentence, self._pending_text

    def _do_lock(self, lock_text, current_words, window_audio, remainder=""):
        """Lock text, slide window, reset agreement state."""
        if self._locked_text:
            self._locked_text = self._locked_text + lock_text
        else:
            self._locked_text = lock_text
        self._locked_sentences.append(lock_text)
        self._pending_text = remainder

        # Slide window
        if len(current_words) > 0:
            locked_count = len(_tokenize(lock_text))
            ratio = locked_count / len(current_words)
            self._window_start_sample += int(ratio * len(window_audio))

        # Reset
        remaining = _tokenize(remainder) if remainder else []
        self._prev_words = remaining
        self._confirmed_words = remaining.copy()
        self._window_pass_count = 0

    def final_transcribe(self, full_audio, language=None):
        """Final pass on remaining audio after last locked sentence.

        Returns:
            (final_remainder: str, full_text: str)
            final_remainder: the last bit of text to paste
            full_text: complete transcription (locked + remainder)
        """
        # Only transcribe audio after locked sentences
        window_audio = full_audio[self._window_start_sample:]

        if len(window_audio) < int(0.3 * self.sr):
            return "", self._locked_text

        text = self.backend.transcribe(window_audio, sr=self.sr, language=language)

        if text.lower().strip() in HALLUCINATIONS:
            text = ""

        # The final remainder to paste
        final_remainder = text.strip()
        if final_remainder and self._locked_text:
            final_remainder = " " + final_remainder

        full_text = self._locked_text + final_remainder

        return final_remainder, full_text

    def get_all_text(self):
        """Return all text produced so far (locked + pending)."""
        if self._pending_text:
            return (self._locked_text + " " + self._pending_text).strip()
        return self._locked_text

    def get_window_duration(self):
        """Current window duration in seconds."""
        return self._window_start_sample / self.sr

    def reset(self):
        """Reset state for next utterance."""
        self._prev_words = []
        self._confirmed_words = []
        self._locked_sentences = []
        self._locked_text = ""
        self._pending_text = ""
        self._window_start_sample = 0
        self._window_pass_count = 0

File: rs/src/streamer.rs
/// Streaming transcription — sentence-based sliding window + LocalAgreement.
///
/// Port of python/engine/streamer.py.
///
/// Strategy:
///   1. Record continuously, transcribe rolling buffer every N seconds
///   2. Word-level LocalAgreement: only trust words that match across 2 runs
///   3. When a complete sentence is confirmed → LOCK it, paste it, slide window
///   4. Re-transcribe only the audio after the locked sentence (~5-10s)
///   5. Final pass: re-transcribe last incomplete sentence for accuracy

use crate::transcriber::Transcriber;

/// Common Whisper hallucinations
const HALLUCINATIONS: &[&str] = &[
    "", "you", "thank you.", "thanks for watching!", "thanks for watching.",
    "subscribe", "bye.", "bye", "thank you", "you.", "the end.",
    "thanks for listening.", "see you next time.", "thank you for watching.", "...",
];

/// Check if char is CJK
fn is_cjk(c: char) -> bool {
    let cp = c as u32;
    (0x4E00..=0x9FFF).contains(&cp)
        || (0x3400..=0x4DBF).contains(&cp)
        || (0x3040..=0x309F).contains(&cp)
        || (0x30A0..=0x30FF).contains(&cp)
        || (0xAC00..=0xD7AF).contains(&cp)
}

/// CJK-aware tokenization (port of Python _tokenize)
fn tokenize(text: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut buf = String::new();
    for ch in text.chars() {
        if is_cjk(ch) {
            if !buf.is_empty() {
                tokens.push(std::mem::take(&mut buf));
            }
            tokens.push(ch.to_string());
        } else if ch.is_whitespace() {
            if !buf.is_empty() {
                tokens.push(std::mem::take(&mut buf));
            }
        } else {
            buf.push(ch);
        }
    }
    if !buf.is_empty() {
        tokens.push(buf);
    }
    tokens
}

/// Join tokens, no space between adjacent CJK (port of Python _join_tokens)
pub fn join_tokens(tokens: &[String]) -> String {
    if tokens.is_empty() {
        return String::new();
    }
    let mut result = tokens[0].clone();
    for i in 1..tokens.len() {
        let prev_cjk = tokens[i - 1].chars().last().map_or(false, is_cjk);
        let cur_cjk = tokens[i].chars().next().map_or(false, is_cjk);
        if !(prev_cjk && cur_cjk) {
            result.push(' ');
        }
        result.push_str(&tokens[i]);
    }
    result
}

/// Longest common prefix of two token lists (port of Python _common_prefix_words)
fn common_prefix_len(a: &[String], b: &[String]) -> usize {
    let len = a.len().min(b.len());
    for i in 0..len {
        if a[i] != b[i] {
            return i;
        }
    }
    len
}

/// Fuzzy agreement: if 80%+ tokens match positionally, return agreed tokens.
/// On mismatch, prefer tokens from words_b (latest run).
/// Port of Python _fuzzy_agree.
fn fuzzy_agree(words_a: &[String], words_b: &[String], threshold: f32) -> (Vec<String>, f32) {
    let length = words_a.len().min(words_b.len());
    if length == 0 {
        return (Vec::new(), 0.0);
    }

    let mut matches = 0;
    let mut agreed = Vec::with_capacity(length);
    for i in 0..length {
        if words_a[i] == words_b[i] {
            matches += 1;
        }
        // Always prefer latest run (words_b) on mismatch
        agreed.push(words_b[i].clone());
    }

    let ratio = matches as f32 / length as f32;
    if ratio >= threshold {
        (agreed, ratio)
    } else {
        (Vec::new(), ratio)
    }
}

/// Check if text ends with punctuation suitable for locking
fn has_lock_punct(text: &str) -> bool {
    text.trim_end()
        .ends_with(|c: char| ".!?。！？,，;；:：、".contains(c))
}

/// Split text into (complete_sentences, remainder) at last punctuation.
/// Port of Python _extract_complete_sentences.
fn extract_complete_sentences(text: &str) -> (String, String) {
    let chars: Vec<char> = text.chars().collect();
    for i in (0..chars.len()).rev() {
        if ".!?。！？,，;；:：、".contains(chars[i]) {
            let complete: String = chars[..=i].iter().collect();
            let remainder: String = chars[i + 1..].iter().collect();
            return (complete.trim().to_string(), remainder.trim().to_string());
        }
    }
    (String::new(), text.to_string())
}

pub struct StreamingTranscriber {
    prev_words: Vec<String>,
    confirmed_words: Vec<String>,
    locked_text: String,
    pending_text: String,
    window_start_sample: usize,
    window_pass_count: usize,
    max_passes: usize,
}

impl StreamingTranscriber {
    pub fn new() -> Self {
        Self {
            prev_words: Vec::new(),
            confirmed_words: Vec::new(),
            locked_text: String::new(),
            pending_text: String::new(),
            window_start_sample: 0,
            window_pass_count: 0,
            max_passes: 3,
        }
    }

    /// Run one rolling transcription pass with fuzzy agreement + fail safe.
    /// Returns (new_sentence, pending_text).
    /// Port of Python rolling_transcribe.
    pub fn rolling_transcribe(
        &mut self,
        window: &[f32],
        transcriber: &Transcriber,
        language: Option<&str>,
    ) -> (Option<String>, String) {
        if window.len() < 8000 {
            // < 0.5s
            return (None, self.pending_text.clone());
        }

        let text = transcriber.transcribe(window, language);

        // Filter hallucinations
        if HALLUCINATIONS.contains(&text.to_lowercase().trim()) {
            return (None, self.pending_text.clone());
        }

        let current_words = tokenize(&text);
        self.window_pass_count += 1;

        // --- Agreement logic (matching Python lines 180-193) ---
        let mut agreed_words: Vec<String> = Vec::new();

        if !self.prev_words.is_empty() {
            // Try strict prefix first (>= 3 tokens)
            let strict_len = common_prefix_len(&self.prev_words, &current_words);
            if strict_len >= 3 {
                agreed_words = self.prev_words[..strict_len].to_vec();
            } else {
                // Try fuzzy agreement (80% threshold)
                let (fuzzy, _ratio) = fuzzy_agree(&self.prev_words, &current_words, 0.8);
                if !fuzzy.is_empty() {
                    agreed_words = fuzzy;
                }
            }
        }

        self.prev_words = current_words.clone();

        // Update confirmed words
        if !agreed_words.is_empty() && agreed_words.len() > self.confirmed_words.len() {
            self.confirmed_words = agreed_words;
        }

        let confirmed_text = join_tokens(&self.confirmed_words);

        // --- Fail safe: force push after N passes on same window ---
        if self.window_pass_count >= self.max_passes {
            let full_text = join_tokens(&current_words);
            if !full_text.trim().is_empty() {
                let (complete, remainder) = extract_complete_sentences(&full_text);
                if !complete.is_empty() {
                    self.do_lock(&complete, &current_words, window.len(), &remainder);
                    return (Some(complete), self.pending_text.clone());
                } else {
                    // No punctuation at all — push everything as last resort
                    self.do_lock(&full_text, &current_words, window.len(), "");
                    return (Some(full_text), self.pending_text.clone());
                }
            }
        }

        // --- Normal lock: agreement + punctuation or length ---
        let mut new_sentence = None;

        if !confirmed_text.is_empty() {
            let (complete, remainder) = extract_complete_sentences(&confirmed_text);

            if !complete.is_empty() {
                new_sentence = Some(complete.clone());
                self.do_lock(&complete, &current_words, window.len(), &remainder);
            } else if confirmed_text.len() > 40 {
                new_sentence = Some(confirmed_text.clone());
                self.do_lock(&confirmed_text, &current_words, window.len(), "");
            } else {
                self.pending_text = confirmed_text;
            }
        } else {
            self.pending_text = join_tokens(&current_words);
        }

        (new_sentence, self.pending_text.clone())
    }

    /// Lock text, slide window, reset agreement state.
    /// Port of Python _do_lock.
    fn do_lock(
        &mut self,
        lock_text: &str,
        current_words: &[String],
        window_len: usize,
        remainder: &str,
    ) {
        self.locked_text.push_str(lock_text);
        self.pending_text = remainder.to_string();

        // Slide window
        if !current_words.is_empty() {
            let locked_count = tokenize(lock_text).len();
            let ratio = locked_count as f32 / current_words.len() as f32;
            self.window_start_sample += (ratio * window_len as f32) as usize;
        }

        // Reset agreement state
        let remaining = if remainder.is_empty() {
            Vec::new()
        } else {
            tokenize(remainder)
        };
        self.prev_words = remaining.clone();
        self.confirmed_words = remaining;
        self.window_pass_count = 0;
    }

    /// Final pass on remaining audio after last locked sentence.
    /// Returns (final_remainder, full_text).
    /// Port of Python final_transcribe.
    pub fn final_transcribe(
        &mut self,
        window: &[f32],
        transcriber: &Transcriber,
        language: Option<&str>,
    ) -> (Option<String>, String) {
        if window.len() < 4800 {
            // < 0.3s
            return (None, self.locked_text.clone());
        }

        let text = transcriber.transcribe(window, language);
        if HALLUCINATIONS.contains(&text.to_lowercase().trim()) {
            return (None, self.locked_text.clone());
        }

        let remainder = text.trim().to_string();
        let mut full = self.locked_text.clone();
        if !remainder.is_empty() {
            if !full.is_empty() {
                full.push(' ');
            }
            full.push_str(&remainder);
        }

        (
            if remainder.is_empty() {
                None
            } else {
                Some(remainder)
            },
            full,
        )
    }

    /// Get full display text (locked + raw current transcription)
    pub fn get_display_text(&self) -> String {
        let raw = join_tokens(&self.prev_words);
        if raw.is_empty() {
            format!("{}{}", self.locked_text, self.pending_text)
        } else {
            format!("{}{}", self.locked_text, raw)
        }
    }

    pub fn confirmed_text(&self) -> String {
        self.locked_text.clone()
    }

    pub fn pending_display_text(&self) -> String {
        let raw = join_tokens(&self.prev_words);
        if raw.is_empty() {
            self.pending_text.clone()
        } else {
            raw
        }
    }

    pub fn window_start_sample(&self) -> usize {
        self.window_start_sample
    }

    pub fn reset(&mut self) {
        self.prev_words.clear();
        self.confirmed_words.clear();
        self.locked_text.clear();
        self.pending_text.clear();
        self.window_start_sample = 0;
        self.window_pass_count = 0;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tokenize_cjk() {
        let tokens = tokenize("今天weather很好");
        assert_eq!(tokens, vec!["今", "天", "weather", "很", "好"]);
    }

    #[test]
    fn test_tokenize_english() {
        let tokens = tokenize("hello world test");
        assert_eq!(tokens, vec!["hello", "world", "test"]);
    }

    #[test]
    fn test_join_tokens_cjk() {
        let tokens: Vec<String> = vec!["今", "天", "weather", "很", "好"]
            .into_iter().map(String::from).collect();
        assert_eq!(join_tokens(&tokens), "今天 weather 很好");
    }

    #[test]
    fn test_fuzzy_agree_exact() {
        let a: Vec<String> = vec!["hello", "world", "test"].into_iter().map(String::from).collect();
        let b = a.clone();
        let (agreed, ratio) = fuzzy_agree(&a, &b, 0.8);
        assert_eq!(agreed.len(), 3);
        assert_eq!(ratio, 1.0);
    }

    #[test]
    fn test_fuzzy_agree_one_diff() {
        let a: Vec<String> = vec!["hello", "world", "test", "here", "now"]
            .into_iter().map(String::from).collect();
        let b: Vec<String> = vec!["hello", "word", "test", "here", "now"]
            .into_iter().map(String::from).collect();
        let (agreed, ratio) = fuzzy_agree(&a, &b, 0.8);
        assert_eq!(agreed.len(), 5); // 4/5 = 80% >= threshold
        assert_eq!(ratio, 0.8);
        assert_eq!(agreed[1], "word"); // prefers latest (b)
    }

    #[test]
    fn test_fuzzy_agree_below_threshold() {
        let a: Vec<String> = vec!["a", "b", "c"].into_iter().map(String::from).collect();
        let b: Vec<String> = vec!["x", "y", "z"].into_iter().map(String::from).collect();
        let (agreed, _) = fuzzy_agree(&a, &b, 0.8);
        assert!(agreed.is_empty());
    }

    #[test]
    fn test_extract_complete_sentences() {
        let (complete, remainder) = extract_complete_sentences("Hello, world. How are you");
        assert_eq!(complete, "Hello, world.");
        assert_eq!(remainder, "How are you");
    }

    #[test]
    fn test_extract_no_punctuation() {
        let (complete, remainder) = extract_complete_sentences("Hello world");
        assert_eq!(complete, "");
        assert_eq!(remainder, "Hello world");
    }
}
